### ISSUE TITLE
refinements of schema format

### ISSUE THREAD

**Issue by lzehl** [openMetadataInitiative/openMINDS_core#131]
Based on discussions with @olinux we're going to simplify the used schema format and provide a generator.py (somewhere) that is going to translate those user-friendly reduced schema formats to proper json-schema (or far future: also other schema formats).

Affected by this is mainly the general implementation of a schema on the Githubs (openMINDS and SANDS), in particular how to reference between schemas. Indirectly this will require an update also on the check performed on the schemas on the Githubs.

NEW schema implementation template:
```json
{
  "_type": "https://openminds.ebrains.eu/SUBDOMAIN/SCHEMALABEL",
  "description": "One sentence description of what metadata this schema covers.",
  "required": [
    "PROPERTYLABEL"
  ],
  "properties": {
    "PROPERTYLABEL": {
      "type": "TYPE",
      "_instruction": "Instruction on what metadata value is expected here."
    },
    "PROPERTYLABEL": {
      "type": "string" OR "array",
      "_instruction": "Enter link to other openMINDS metadata file of the stated schema type.",
      "_linkedTypes": [
        "https://openminds.ebrains.eu/SUBDOMAIN/SCHEMALABEL"
      ]
    }
  }
}
```

Further instructions / explanations:
- **`"_type"`** will replace (combine) the previous **`"$id"`** and **`"@type"`**.
- **SUBDOMAIN options** (with incomplete tier 3 set):
  - `"core"` (tier 1)
  - `"sands"` (tier 2)
  - `"electrophysiology"` (tier 3)
  - ??? (tier 3)
- **SCHEMALABEL** will be capitalized (example: **`"ResearchProduct"`**).
- **`"_linkedTypes"`** will replace the previous **`"$ref"`**
  - if a single link is expected for a **`"_linkedTypes"`** property use `"type": "string"`
  - if multiple links are expected for a **`"_linkedTypes"`** property use `"type": "array"`
  - in case of **`"_linkedTypes"`**, additional keys for properties of `"type": "array"` are the same as in json-schema without defining **`"items"`**
  - note: within an array items have to be of the same type (schema types listed in **`"_linkedTypes"`** means OR only)

Note: A definitions.json is not gonna be needed any more after this change. Also not needed anymore for the schema implementation is the **`"@id"`** and **`"@type"`** (they will be taken care of "globally" by the generator).

Note for controlledTerms: The generic reference will be `"_linkedTypes": "https://openminds.ebrains.eu/controlledTerms"`. To refer to specific types of controlledTerms use `"_linkedTypes": "https://openminds.ebrains.eu/controlledTerms/TERMTYPE"`. TERMTYPE could be, e.g. BiologicalSex, Laterality, etc

**Comment by lzehl** [comment]
@olinux please verify the documentation I provided above. Thank you :slightly_smiling_face:

**Comment by lzehl** [comment]
@olinux in SANDS @xgui3783 brought up his concern about using the same prefix for the technical keys as json-schema ("$"). I don't have a strong opinion here. If you agree with @xgui3783 that it might be less confusing we could define our own prefix (I just would prefer something very simple, e.g. "_" or "#" etc).

**Comment by olinux** [comment]
> @olinux in SANDS @xgui3783 brought up his concern about using the same prefix for the technical keys as json-schema ("$"). I don't have a strong opinion here. If you agree with @xgui3783 that it might be less confusing we could define our own prefix (I just would prefer something very simple, e.g. "_" or "#" etc).

I see - I don't have a strong opinion either. We could put some emphasis on being a European project by using € instead ;)

**Comment by lzehl** [comment]
:smile: I had the same thought.... would that cause any issues (formatting wise in python for example)?

**Comment by olinux** [comment]
> smile I had the same thought.... would that cause any issues (formatting wise in python for example)?

not as long as we're sticking with at least utf-8

**Comment by lzehl** [comment]
@xgui3783 would switching to "€" as prefix satisfy your concern?

Additional note: the validator on the Githubs needs to be adopted to the new "EBRAINS" schema format soon, but for now it's just reduced to check for a valid json so that we can continue the implementation.

**Comment by xgui3783** [comment]
> would switching to "€" as prefix satisfy your concern?

Yes!

T.T Though I have no "€" sign on my keyboard. (It's a small annoyance, but I am sure I can get used to it. But I think this may send the wrong message to international neuroscience communities)

Re: proposal, am I understanding correctly that the main benefit of introducing the schema is to reduce verbosity? The additional work we add to ourselves is not insignificant (writing and maintaining ebrains schema, maintaining generator.py, supporting new json-schema versions).

**Comment by lzehl** [comment]
@xgui3783 not only. Generally we just have the problem that neither json-schema nor shacl are fulfilling our needs. Meaning that the additional work of a generator which compensates for the shortcomings anyway will not be avoidable. By setting up our "own" schema language (which is much less technical and therefore much more user-friendly for less experienced programmers) we can be more steady in our development, and can flexibly transform it in all the schema-languages we see fit / useful (for example this includes for now the json-schema and the html representation, but in future maybe also shacl or odML). This is why I think the solution is very elegant and worth the maintenance effort which we cannot avoid in any case.

**Comment by lzehl** [comment]
Other prefix suggestions are welcome! As I said: I don't have a strong opinion here, it should just be very simple.

**Comment by xgui3783** [comment]
Aw, ok I understand. I have clarified my stance also on sands thread https://github.com/HumanBrainProject/SANDS/pull/92#issuecomment-721068861

> Other prefix suggestions are welcome! As I said: I don't have a strong opinion here, it should just be very simple.

I mean, we could keep it officially as the euro sign, but have a possible alias (like "__", so "€type" could also be "__type")? or would that complicate things?

I would very much like it if the symbol can be widely accessible.

**Comment by lzehl** [comment]
@olinux & @xgui3783 the concern you have with "€" is valid. Let's go for now with "_".

**Comment by lzehl** [comment]
@olinux three questions:

1) could we introduce a "_label" (see above) this would match the "title" of json-schema.
2) should we still use the file extension "....schema.json"? we are not really json-schema conform in these files and it might be confusing for users.
3) please have a look again at the "_linkedTypes" properties; what does the user need to provide here locally? based on our previous discussions, we said the user should provide a string (filepath to correponding json-LD, or KG ID of corresponding metadata instance); the generator/compiler needs to solve this then into a valid object-link; did I misunderstood this?

**Comment by UlrikeS91** [comment]
> 1. could we introduce a "_label" (see above) this would match the "title" of json-schema.

@olinux and I talked about this earlier today, and would like to discuss this more. If we were to use it, @olinux would prefer to stick to `title`, but he also had very good arguments to not use it at all. @olinux, could you please give a quick summary of your view point?

**Comment by UlrikeS91** [comment]
Additionally, @olinux suggested to replace `instruction` by `_instruction`. Since this doesn't exist in json-schema itself, we should be consistent in adding the underscore when it is something we invented.

**Comment by lzehl** [comment]
> Additionally, @olinux suggested to replace `instruction` by `_instruction`. Since this doesn't exist in json-schema itself, we should be consistent in adding the underscore when it is something we invented.

I'm fine with that. Makes total sense.

**Comment by lzehl** [comment]
> > 1. could we introduce a "_label" (see above) this would match the "title" of json-schema.
> 
> @olinux and I talked about this earlier today, and would like to discuss this more. If we were to use it, @olinux would prefer to stick to `title`, but he also had very good arguments to not use it at all. @olinux, could you please give a quick summary of your view point?

I suggest to leave it out then. It can be easily inferred by "_type"

**Comment by lzehl** [comment]
I've updated the documentation of the new schema layout above, according to your comments. 
@UlrikeS91 & @olinux What about point 3? Or should this one be better discussed in person?

**Comment by lzehl** [comment]
solved with #138
